code/acting/src/acting/pure_pursuit_controller.py

```diff
@@ -0,0 +1,305 @@
+            self.logwarn("__get_heading: Last Position not set")
+            self.__heading = 0
+            return
+
+        cur_x, cur_y = points_to_vector(
+                                        (self.__last_pos[0],
+                                         self.__last_pos[1]),
+                                        (self.__position[0],
+                                         self.__position[1])
+                                        )
+        # maybe remove weight if it doesn't help (after fixing gps signal)
+        # code without weight:
+        # self.__heading = vectors_to_angle(cur_x, cur_y, 1, 0)
+        # ->
+        if self.__heading is not None:
+            old_heading: float = self.__heading
+            new_heading: float = vectors_to_angle(cur_x, cur_y, 1, 0)
+            self.__heading = (2 * new_heading + 1 * old_heading) / 3
+        else:
+            self.__heading = vectors_to_angle(cur_x, cur_y, 1, 0)
+        # <-
+
+    def __set_velocity(self, data: CarlaSpeedometer):
+        self.__velocity = data.speed
+
+    def __calculate_steer(self) -> float:
+        """
+        Calculates the steering angle based on the current information
+        :return:
+        """
+        l_vehicle = 2.85  # wheelbase
+        k_ld = 7.5  # todo: tune
+
+        current_velocity: float
+        if self.__velocity == 0:
+            current_velocity = k_ld * 0.1
+        else:
+            current_velocity = self.__velocity
+
+        look_ahead_dist = k_ld * current_velocity
+        self.__tp_idx = self.__get_target_point_index(look_ahead_dist)
+
+        target_wp: PoseStamped = self.__path.poses[self.__tp_idx]
+
+        target_v_x, target_v_y = points_to_vector((self.__last_pos[0],
+                                                   self.__last_pos[1]),
+                                                  (target_wp.pose.position.x,
+                                                   target_wp.pose.position.y))
+
+        zero_h_v_x, zero_h_v_y = points_to_vector((self.__last_pos[0],
+                                                   self.__last_pos[1]),
+                                                  (self.__last_pos[0] + 1,
+                                                   self.__last_pos[1]))
+
+        target_vector_heading = vectors_to_angle(target_v_x, target_v_y,
+                                                 zero_h_v_x, zero_h_v_y)
+
+        alpha = self.__heading - target_vector_heading
+        steering_angle = atan((2 * l_vehicle * sin(math.radians(alpha))) /
+                              look_ahead_dist)
+
+        # for debugging ->
+        debug_msg = Debug()
+        debug_msg.heading = self.__heading
+        debug_msg.target_heading = target_vector_heading
+        debug_msg.l_distance = look_ahead_dist
+        debug_msg.alpha = alpha
+        self.debug_publisher.publish(debug_msg)
+        # <-
+
+        self.pure_pursuit_steer_target_pub.publish(target_wp.pose)
```

Comment:
Is there a reason you use the `pose` but not the `poseStamped`?

Reply:
I did not see the need for the added header. If this turns out to be a problem, it would be a quick fix later on.

Reply:
Alright, just have been curious